earth2studio/utils/interp.py

```diff
@@ -0,0 +1,170 @@
+
+    lat1, lon1 are assumed to be 2-dimensional meshgrids with possibly unequal spacing.
+
+    Parameters
+    ----------
+    values : torch.Tensor [..., H_in, W_in]
+        Input values defined over (lat0, lon0) that will be interpolated onto
+        (lat1, lon1) grid.
+    lat0 : torch.Tensor [W_in, ]
+        Vector of input latitude coordinates, assumed to be increasing with
+        equal spacing.
+    lon0 : torch.Tensor [W_in, ]
+        Vector of input longitude coordinates, assumed to be increasing with
+        equal spacing.
+    lat1 : torch.Tensor [H_out, W_out]
+        Tensor of output latitude coordinates
+    lon1 : torch.Tensor [H_out, W_out]
+        Tensor of output longitude coordinates
+
+    Returns
+    -------
+    result : torch.Tensor [..., H_out, W_out]
+        Tensor of interpolated values onto lat1, lon1 grid.
+    """
+
+    # Get input grid shape and flatten
+    latshape, lonshape = lat1.shape
+    lat1 = lat1.flatten()
+    lon1 = lon1.flatten()
+
+    # Get indices of nearest points
+    latinds = torch.searchsorted(lat0, lat1) - 1
+    loninds = torch.searchsorted(lon0, lon1) - 1
+
+    # Get original grid spacing
+    dlat = lat0[1] - lat0[0]
+    dlon = lon0[1] - lon0[0]
+
+    # Get unit distances
+    normed_lat_distance = (lat1 - lat0[latinds]) / dlat
+    normed_lon_distance = (lon1 - lon0[loninds]) / dlon
+
+    # Apply bilinear mapping
+    result = (
+        values[..., latinds, loninds]
+        * (1 - normed_lat_distance)
+        * (1 - normed_lon_distance)
+    )
+    result += (
+        values[..., latinds, loninds + 1]
+        * (1 - normed_lat_distance)
+        * (normed_lon_distance)
+    )
+    result += (
+        values[..., latinds + 1, loninds]
+        * (normed_lat_distance)
+        * (1 - normed_lon_distance)
+    )
+    result += (
+        values[..., latinds + 1, loninds + 1]
+        * (normed_lat_distance)
+        * (normed_lon_distance)
+    )
+    return result.reshape(*values.shape[:-2], latshape, lonshape)
+
+
+class LatLonInterpolation(nn.Module):
+    """Bilinear interpolation between arbitrary grids.
+
+    The mapped values can be on arbitrary grid, but the output grid should be
+    contained within the input grid.
+
+    Initializing the interpolation object can be somewhat slow, but interpolation is
+    fast and runs on the GPU once initialized. Therefore, prefer to reuse the
+    interpolation object when possible.
+
+    Parameters
+    ----------
+    lat0 : torch.Tensor [H_in, W_in]
+        Tensor of input latitude coordinates
+    lon0 : torch.Tensor [H_in, W_in]
+        Tensor of input longitude coordinates
+    lat1 : torch.Tensor [W_out, H_out]
+        Tensor of output latitude coordinates
+    lon1 : torch.Tensor [W_out, H_out]
+        Tensor of output longitude coordinates
+    """
+
+    def __init__(
+        self,
+        lat_in: Tensor | ArrayLike,
+        lon_in: Tensor | ArrayLike,
```

Comment:
Nit: typically like to use torch.Tensor here just so its explicit what tensor/package its expecting

Reply:
Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/cc59471fae79f5305c9e0cb512ca4d56e0f76834.